Implement basic Parquet data file reading capability

sdd opened the pull request:
A refactor of @ZENOTME 's PR (https://github.com/apache/iceberg-rust/pull/199) based on some of the comments on that PR. Working and with a test. Completes https://github.com/apache/iceberg-rust/issues/125

EDIT: refactored. inspired by the ArrowReader in the Java implementation. TODO: schema translation

liurenjie1024 commented on crates/iceberg/src/spec/manifest.rs:
```diff
@@ -855,6 +855,11 @@ impl ManifestEntry {
     }
 
+    /// Content type of this manifest entry.
+    pub fn file_format(&self) -> DataFileFormat {
```
```suggestion
    #[inline(always)]
    pub fn file_format(&self) -> DataFileFormat {
```

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -180,7 +185,42 @@ pub type ArrowRecordBatchStream = BoxStream<'static, crate::Result<RecordBatch>>
     /// Returns a stream of arrow record batches.
     pub async fn execute(&self) -> crate::Result<ArrowRecordBatchStream> {
-        todo!()
+        match self.data_file.content_type() {
```
After checking java/python's implementation, I think it's in appropriate to have this api. To convert a `FileScanTask` to arrow record batch stream, we may need some other configurations:
1. Record batch size, e.g. number of records of each record batch, this may affect memory usage of upstream executors
2. `FileIO`, as in your implementation.

Also it's more general to have multi `FileScanTask`s in size based reader.

ZENOTME replied:
And seems it should not contain `FileIO` in FileScanTask if we want it can be de/serde. Maybe we can get a stream of FileScanTask and use FileIO and FileScanTask to construct the reader.

sdd replied:
FileScanTask no longer has ref to fileio in latest version

liurenjie1024 reviewed (commented):
Hi, @sdd Thanks for the effort, it looks great! But there are some things I think it would be better to do:
1. We can have sth like [`ArrowReader`](https://github.com/apache/iceberg/blob/97a99c3cf4f3293fe3536402f015c5b65906e0ba/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/ArrowReader.java#L101) as in java api, which accepts some arguments for reading, e.g. a stream of `FileScanTask`, configuration like record batch size.
2. We need to add some conversion from arrow schema to iceberg schema, since it determines position of each array.

liurenjie1024 commented:
The process of reading into arrow record batch stream seems more complex than I expected, you can refer [the codes in python](https://github.com/apache/iceberg-python/blob/6c8ea0effac0942ad4e880e5eef627473a354040/pyiceberg/io/pyarrow.py#L788). I'm planning to setup a framework for this so that contributors can help to fill in more details. cc @sdd What do you think?

sdd commented:
Hi @liurenjie1024 - Thanks for the review. I agree with your conclusions. I'll take a closer look at the Java and Python implementations. The framework approach sounds sensible - happy to work with you and others to get this into the shape it needs to be 👍🏼

liurenjie1024 commented:
> Thanks for those links, Renjie, they were very useful.. I've taken a deeper look at `ArrowReader` and the python code. I'll submit an improved design here tomorrow morning (UK) with some `todo!`s as subtasks so that others can work on this too.

Really cool!

sdd commented:
It's taken me a bit longer than my morning train journey to get the design ready. Will continue on the way home later 😁

sdd commented:
PTAL @liurenjie1024 @ZENOTME - took a different approach, inspired by the Java implementation but a lot simpler.

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -178,9 +241,16 @@ pub struct FileScanTask {
-    /// Returns a stream of arrow record batches.
-    pub async fn execute(&self) -> crate::Result<ArrowRecordBatchStream> {
-        todo!()
+    pub fn data_file(&self) -> ManifestEntryRef {
+        self.data_file.clone()
+    }
+
+    pub fn start(&self) -> u64 {
+        self.start
+    }
+
+    pub fn length(&self) -> u64 {
+        self.length
```
Currently these fields are only used to by `open` methods, so I think it's unnecessary to expose them?

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+        todo!()
+    }
+
+    fn get_arrow_row_selection(&self) -> RowSelection {
```
We can ignore this for now as it's non trivial to do it.

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+        )
+    }
+
+    fn get_arrow_projection_mask(&self) -> ProjectionMask {
```
In fact, we need to calculate the `ProjectionMask` for each parquet file due to schema evolution of iceberg, the schema projection of each parquet file may be different.

sdd replied:
Aah, of course. Will update

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+                            Error::new(ErrorKind::Unexpected, "failed to load parquet file").with_source(err)
+                        })?
+                        .with_batch_size(batch_size)
+                        .with_offset(task.start() as usize)
+                        .with_limit(task.length() as usize)
```
I think we should omit these two for now.  The `start`/`length` was designed to be file pos, while `offset`/`limit` was for row number. We need a conversion for these, but they can be left for later.

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+                    let mut batch_stream = ParquetRecordBatchStreamBuilder::new(parquet_reader)
+                        .await
+                        .map_err(|err| {
+                            Error::new(ErrorKind::Unexpected, "failed to load parquet file").with_source(err)
```
```suggestion
                            Error::new(ErrorKind::Unexpected, "Failed to load parquet file").with_source(err).with_context("filename", "<input file path>")
```
Please also add context here better error reporting.

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+                        .with_projection(projection_mask.clone())
+                        .with_row_selection(row_selection.clone())
+                        .build()
+                        .unwrap()
```
Why we panic here?

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,6 +178,54 @@ impl TableScan {
+
+    /// Transforms a stream of FileScanTasks from plan_files into a stream of
+    /// Arrow RecordBatches.
+    pub fn open(&self, mut tasks: FileScanTaskStream) -> crate::Result<ArrowRecordBatchStream> {
```
This method body's implementation is correct for one parquet file. For a stream of input files, we can use [futures::StreamExt::flat_map](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.flat_map) to combine them into a  stream of record batches.

sdd replied:
> Putting `open` in `TableScan` seems weird to me. The `TableScan` has two responsibilities: generate the task and open the task and it will make it complex. Can we separate it out to the type, e.g. xxReader.🤔

Sure, I can do that.

liurenjie1024 reviewed (commented):
@sdd Thanks for this. I left some comments to further simplify some unnecessary changes.

sdd commented:
@liurenjie1024 and @ZENOTME : PTAL again, I've refactored the parquet file reading logic out into a separate reader struct. I've addressed all the other comments also.

ZENOTME reviewed (approved):
Thanks! Nice job! @sdd

ZENOTME commented on crates/iceberg/src/io.rs:
```diff
@@ -215,9 +216,12 @@ pub struct InputFile {
 
 /// Trait for reading file.
-pub trait FileRead: AsyncRead + AsyncSeek {}
+pub trait FileRead: AsyncRead + AsyncSeek + Send + Unpin + TokioAsyncRead + TokioAsyncSeek {}
```
> TODO: I'm working on https://github.com/upondal/opendal-parquet-io for best performance.

Maybe we can use an interface like `InputFile.parquet_reader() -> AsyncFIleReader` to let a FileIO provide a custom parquet reader. But for now, using the FileRead trait looks good to me. cc @Xuanwo

Xuanwo replied:
I agree. Let's move first.

liurenjie1024 commented on crates/iceberg/src/file_record_batch_reader.rs:
Given that most of these operations are arrow related, how about rename this file into `arrow.rs`?  We will also put arrow schema conversion here.

liurenjie1024 commented on crates/iceberg/src/file_record_batch_reader.rs:
```diff
@@ -0,0 +1,83 @@
+
+impl FileRecordBatchReader {
+    /// Constructs a new FileRecordBatchReader
+    pub fn new(file_io: FileIO, schema: SchemaRef, batch_size: Option<usize>) -> Self {
```
How about replacing this with builder pattern?

sdd replied:
I can if you really think that's best, but that feels quite heavy when there is only one parameter that is optional?

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,24 +164,32 @@ impl TableScan {
         Ok(iter(file_scan_tasks).boxed())
     }
+
+    pub async fn execute(
```
```suggestion
    pub async fn to_arrow(
```

How about making things more explicit ?

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -163,24 +164,32 @@ impl TableScan {
+
+    pub async fn execute(
+        &self,
+        batch_size: Option<usize>,
```
This should be put in `TableScanBuilder` rather than just passing as function argument.

sdd replied:
I can do this, but I wasn't sure that made sense, since `batch_size` might be only relevant for `TableScanBuilder`'s whose `to_arrow()` method get called? Will we have other methods on here equivalent to `to_arrow()` but returning different types?

liurenjie1024 replied:
Yes, you are right. `TableScan` is used to define scan related args, such as filter, snapshot, while `batch_size` is mostly execution related, so it's appropriate to associate it with `ArrowReader`. The problem is with this api is that it's not extensible, since we need to break it every time we add a new config option for `ArrowReader`. How about removing this api, but passing `TableScan` to `ArrowReaderBuilder`?

sdd replied:
I've done this now

liurenjie1024 commented on crates/iceberg/src/scan.rs:
```diff
@@ -445,4 +494,22 @@ mod tests {
     }
+
+    #[tokio::test]
+    async fn test_open_parquet_no_deletions() {
```
While this is currently fine for unit tests, it would be better to add integrations for this.

sdd replied:
Thanks - I'll create an integration test in the follow-up PRs that extend the functionality

liurenjie1024 reviewed (commented):
@sdd Thanks for this, we are quite close! Left some minor suggestions to improve.

sdd commented:
Ok @liurenjie1024 @Xuanwo and @ZENOTME, I've addressed the feedback and it's ready to review again. Thanks!

liurenjie1024 reviewed (approved):
LGTM, thanks!

Xuanwo reviewed (approved):
Thanks!

sdd commented:
@Fokko are you able to take a look at this one yet please? 😁
